pkg/fleets/controller.go

```diff
@@ -173,39 +224,102 @@ func (c *Controller) syncFleet(key string) error {
 		return err
 	}
 
-	var activeGsSet *stablev1alpha1.GameServerSet
+	activeGsSet, rest := c.filterGameServerSetByActive(fleet, list)
+	if err := c.applyDeploymentStrategy(fleet, rest); err != nil {
+		return err
+	}
+	if err := c.deleteEmptyGameServerSets(fleet, rest); err != nil {
+		return err
+	}
 
-	// if there isn't a GameServerSet yet, then create one
-	if len(list) == 0 {
+	// if there isn't an active gameServerSet, create one (but don't persist yet)
+	if activeGsSet == nil {
+		c.logger.WithField("fleet", fleet.ObjectMeta.Name).Info("could not find active GameServerSet, creating")
 		activeGsSet = fleet.GameServerSet()
-		activeGsSet, err = c.gameServerSetGetter.GameServerSets(fleet.ObjectMeta.Namespace).Create(activeGsSet)
+	}
+
+	replicas := fleet.ReplicasMinusSumAllocated(rest)
+	if err := c.upsertGameServerSet(fleet, activeGsSet, replicas); err != nil {
+		return err
+	}
+	return c.updateFleetStatus(fleet)
+}
+
+// upsertGameServerSet if the GameServerSet is new, insert it
+// if the replicas minus sum allocated does not match the active
+// GameServerSet, then update it
+func (c *Controller) upsertGameServerSet(fleet *stablev1alpha1.Fleet, gsSet *stablev1alpha1.GameServerSet, replicas int32) error {
+	if gsSet.ObjectMeta.UID == "" {
+		gsSet.Spec.Replicas = replicas
+		gsSet, err := c.gameServerSetGetter.GameServerSets(gsSet.ObjectMeta.Namespace).Create(gsSet)
 		if err != nil {
 			return errors.Wrapf(err, "error creating gameserverset for fleet %s", fleet.ObjectMeta.Name)
 		}
 
 		c.recorder.Eventf(fleet, corev1.EventTypeNormal, "CreatingGameServerSet",
-			"Created GameServerSet %s", activeGsSet.ObjectMeta.Name)
+			"Created GameServerSet %s", gsSet.ObjectMeta.Name)
+	} else if replicas != gsSet.Spec.Replicas {
```

Comment:
Rather than use an `else if` statement here, you could add a `return nil` to the `if` block above, and a `return nil` to the `else if` block. That way the control flow is easier to follow and more idiomatic.

Reply:
Updated. That should be better. Definitely seems more readable :+1: